flight/Modules/VtolPathFollower/vtol_follower_fsm.c

```diff
@@ -255,18 +302,55 @@ const static struct vtol_fsm_transition *current_goal;
 static enum vtol_fsm_state curr_state;
 
 /**
- * Process any sequence of automatic state transitions
+ * Enter a new state.  Reacts appropriately (e.g. does nothing) if you pass
+ * FSM_STATE_UNCHANGED.
+ * @param[in] state The desired state.
+ * @return true if there was a state transition, false otherwise.
  */
-static void vtol_fsm_process_auto()
+static int vtol_fsm_enter_state(enum vtol_fsm_state state)
 {
-	while (current_goal[curr_state].next_state[FSM_EVENT_AUTO]) {
-		curr_state = current_goal[curr_state].next_state[FSM_EVENT_AUTO];
+	if (state != FSM_STATE_UNCHANGED) {
+		curr_state = state;
 
 		/* Call the entry function (if any) for the next state. */
 		if (current_goal[curr_state].entry_fn) {
 			current_goal[curr_state].entry_fn();
 		}
+
+		/* 0 disables any pending timeout, otherwise it's set to the
+		 * value for this state */
+		vtol_fsm_timer_set(current_goal[curr_state].timeout);
```

Comment:
ah, my thought was if the entry_fn was time consuming and bound up the fsm thread, it should get a chance to do that before the timer gets set.  Willing to move up or even to do something fancy (set 0, then if still 0 set to prespecified timeout), just lmk.